pkg/cmd/pulumi/templatecmd/template_publish.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/archive"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+)
+
+type publishTemplateArgs struct {
+	publisher string
+	version   string
+	name      string
+}
+
+type templatePublishCmd struct {
+	defaultOrg func(context.Context, backend.Backend, *workspace.Project) (string, error)
+}
+
+func newTemplatePublishCmd() *cobra.Command {
+	args := publishTemplateArgs{}
+	var tplPublishCmd templatePublishCmd
+
+	cmd := &cobra.Command{
+		Use:   "publish <directory>",
+		Args:  cmdutil.ExactArgs(1),
+		Short: "Publish a template to the Pulumi Registry",
+		Long: "Publish a template to the Pulumi Registry.\n\n" +
+			"This command publishes a template directory to the Pulumi Registry. " +
+			"The directory will be archived as a .tar.gz file and uploaded.",
+		RunE: func(cmd *cobra.Command, cliArgs []string) error {
+			ctx := cmd.Context()
+			tplPublishCmd.defaultOrg = backend.GetDefaultOrg
+			return tplPublishCmd.Run(ctx, args, cliArgs[0])
+		},
+	}
+
+	cmd.Flags().StringVar(
+		&args.version, "version", "",
+		"The version of the template (required, semver format)")
+	cmd.Flags().StringVar(
+		&args.name, "name", "",
+		"The name of the template (required)")
+	cmd.Flags().StringVar(
+		&args.publisher, "publisher", "",
+		"The publisher of the template (e.g., 'pulumi'). Defaults to the default organization in your pulumi config.")
+	err := cmd.MarkFlagRequired("version")
+	if err != nil {
+		panic(err)
+	}
+	err = cmd.MarkFlagRequired("name")
+	if err != nil {
+		panic(err)
+	}
+
+	return cmd
+}
+
+func (cmd *templatePublishCmd) Run(
+	ctx context.Context,
+	args publishTemplateArgs,
+	templateDir string,
+) error {
+	if _, err := os.Stat(templateDir); os.IsNotExist(err) {
+		return fmt.Errorf("template directory does not exist: %s", templateDir)
+	}
+
+	if args.version == "" {
+		return errors.New("version is required, please specify --version")
+	}
+	version, err := semver.ParseTolerant(args.version)
+	if err != nil {
+		return fmt.Errorf("invalid version format: %w", err)
+	}
+
+	project, _, err := pkgWorkspace.Instance.ReadProject()
+	if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+		return fmt.Errorf("failed to determine current project: %w", err)
+	}
+
+	b, err := login(ctx, project)
+	if err != nil {
+		return err
+	}
+
+	var publisher string
+	if args.publisher != "" {
+		publisher = args.publisher
+	} else {
+		publisher, err = cmd.defaultOrg(ctx, b, project)
+		if err != nil {
+			return fmt.Errorf("failed to determine default organization: %w", err)
+		}
+		if publisher == "" {
+			return errors.New(
+				"no publisher specified and no default organization found, " +
+					"please set a publisher or set a default organization in your pulumi config",
+			)
+		}
+	}
+
+	fmt.Printf("Creating archive from directory: %s\n", templateDir)
```

Comment:
To keep this easy to test, let's use cobra's `cmd.ErrOrStderr()`.
```suggestion
	fmt.Fprintf(cmd.ErrOrStderr(), "Creating archive from directory: %s\n", templateDir)
```

Likewise for any other `fmt.Printf` in the command.